### ISSUE TITLE
SDPA-decode: rel_l2 ~0.026 against fp32 reference on bf16 inputs, not visible to PCC

### ISSUE THREAD

**Issue by dgolubovicTT** [tenstorrent/tt-metal#51927]
Caught while debugging this [issue](https://github.com/tenstorrent/tt-inference-server/issues/4752)

## Repro

```
python tests/ttnn/unit_tests/operations/sdpa/paged_sdpa_decode_accuracy_repro.py
```

Script at the bottom (self-contained, needs only `ttnn` + `torch`). Blackhole, `main` @ `82ef8703abb`, grid 11x10. The reference is fp32 torch computed from **the same bf16 values written to device**, so input rounding is excluded.

## Observed

`paged_scaled_dot_product_attention_decode`, batch 32, Hq 12, Hkv 4, head_dim 256, paged K/V `[1024, 4, 32, 256]` bf16, `scale=0.0625`, `cur_pos ~ U[1,256)`, `rel_l2 = ||device - ref||_2 / ||ref||_2`:

| operands | rel_l2 | PCC |
|---|---|---|
| iid gaussian Q/K/V, 10 seeds | median **0.0255** (0.0225–0.0279) | 0.99970–0.99981 |

- PCC stays above 0.9996 throughout, so PCC-gated tests pass.
- Flat in sequence length — 32→1024 tokens gives 0.0265→0.0284, while bf16 accumulation noise (`sqrt(N)*eps`) over that range would grow 0.022→0.125.
- Same numbers for the non-paged `scaled_dot_product_attention_decode`.
- `fp32_dest_acc_en=True` gives median 0.0203 (improves on 10/10 seeds).

## Additional observation: the single-key case is exactly wrong

With `cur_pos = 0` exactly one key is unmasked, so softmax over one element is 1.0 and the output must equal `V[0]`. It is instead `0.9702 * V[0]` — uniformly, elementwise. Dividing that one factor out leaves rel_l2 **0.0023**, i.e. bf16 round-off, so in this degenerate case the factor is the entire error.

That factor looks independent of the data rather than a property of one sample: it is `0.9702 ± 0.0006` across all 384 individual (user, head) pairs, each with independently drawn Q/K/V, and `0.970164 ± 0.00002` across batch 1/32, Hq 12/32, Hkv 4/8/32, `k_chunk_size` 32/64/128, `exp_approx_mode` True/False, and paged vs non-paged.

**It does not, however, explain the error at realistic decode positions.**

Separately: `exp_approx_mode` produced bit-identical output in every case above.

<details>
<summary>repro script</summary>

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
# SPDX-License-Identifier: Apache-2.0

"""Repro: SDPA-decode applies a systematic ~0.9702 scale factor to its output.

Self-contained; needs only ttnn + torch:

    python tests/ttnn/unit_tests/operations/sdpa/paged_sdpa_decode_accuracy_repro.py

Three sections:

1. self-check    - MHA sanity gate, so a mis-specified operand cannot masquerade
                   as a kernel defect.
2. single token  - the minimal repro. With ``cur_pos = 0`` exactly one key is
                   unmasked, so softmax over one element is exactly 1.0 and the
                   output must be a bit-exact copy of V[0]. It is instead
                   0.970164 * V[0]. Dividing that one factor out leaves rel_l2
                   0.0023, i.e. plain bf16 round-off, so the scale factor is
                   essentially the entire error.
3. 10 seeds      - the same defect at realistic decode positions, with and
                   without ``fp32_dest_acc_en``.

Metric is rel_l2 = ||device - reference||_2 / ||reference||_2, not PCC: PCC is
scale-invariant by construction, so a purely multiplicative error is precisely
what it cannot see. PCC stays above 0.9997 everywhere below.

References are computed in fp32 from the *same bf16 values written to device*, so
input rounding is excluded and what is measured is the kernel's own error.
"""

import argparse

import torch

import ttnn

BATCH = 32
NUM_Q_HEADS = 12
NUM_KV_HEADS = 4
HEAD_DIM = 256
BLOCK_SIZE = 32
BLOCKS_PER_USER = 32
MAX_SEQ_LEN = BLOCK_SIZE * BLOCKS_PER_USER  # 1024
SCALE = HEAD_DIM**-0.5  # 0.0625


def torch_reference(q, k_per_user, v_per_user, cur_pos, scale):
    """Causal decode SDPA in fp32. cur_pos[b] is user b's last token, inclusive."""
    b_sz, n_q, n_kv = q.shape[1], q.shape[2], k_per_user.shape[1]
    repeat = n_q // n_kv
    out = torch.zeros(1, b_sz, n_q, v_per_user.shape[3], dtype=torch.float32)
    for b in range(b_sz):
        pos = int(cur_pos[b].item())
        k_b = k_per_user[b, :, : pos + 1, :].float().repeat_interleave(repeat, dim=0)
        v_b = v_per_user[b, :, : pos + 1, :].float().repeat_interleave(repeat, dim=0)
        scores = torch.einsum("hd,hsd->hs", q[0, b].float(), k_b) * scale
        out[0, b] = torch.einsum("hs,hsd->hd", torch.softmax(scores, dim=-1), v_b)
    return out


def paged_layout(per_user, page_table, num_blocks):
    """Scatter per-user K/V into a paged buffer via page_table[user, vblock] -> pblock."""
    paged = torch.zeros(
        num_blocks, per_user.shape[1], BLOCK_SIZE, per_user.shape[3], dtype=per_user.dtype
    )
    for b in range(per_user.shape[0]):
        for vb in range(page_table.shape[1]):
            paged[int(page_table[b, vb].item())] = per_user[
                b, :, vb * BLOCK_SIZE : (vb + 1) * BLOCK_SIZE, :
            ]
    return paged


def rel_l2(got, want):
    return (
        torch.linalg.vector_norm((got.float() - want.float()))
        / torch.linalg.vector_norm(want.float())
    ).item()


def pcc(a, b):
    x, y = a.float().flatten(), b.float().flatten()
    x, y = x - x.mean(), y - y.mean()
    return (torch.dot(x, y) / (x.norm() * y.norm())).item()


def build(device, seed, batch, n_q, n_kv, pos_fn, k_chunk=32, exp_approx=False):
    """Returns (device output sliced to n_q heads, q, k_per_user, v_per_user, cur_pos)."""
    torch.manual_seed(seed)
    num_blocks = batch * BLOCKS_PER_USER
    # bf16 round-trip before the reference sees them.
    q = torch.randn(1, batch, n_q, HEAD_DIM).bfloat16().float()
    k = torch.randn(batch, n_kv, MAX_SEQ_LEN, HEAD_DIM).bfloat16().float()
    v = torch.randn(batch, n_kv, MAX_SEQ_LEN, HEAD_DIM).bfloat16().float()
    # One distinct physical block per (user, slot): no user aliases another's blocks.
    page_table = torch.arange(num_blocks, dtype=torch.int32).reshape(batch, BLOCKS_PER_USER)
    cur_pos = pos_fn(batch)

    # Keep the logical q-head count at n_q and let TILE_LAYOUT pad internally.
    # Hand-padding q to a full 32-head tile would tell the op the GQA ratio is
    # 32/n_kv instead of n_q/n_kv, silently remapping q heads to kv heads.
    tt = lambda t, layout=True: ttnn.from_torch(  # noqa: E731
        t,
        dtype=ttnn.bfloat16 if t.is_floating_point() else ttnn.int32,
        layout=ttnn.TILE_LAYOUT if layout else ttnn.ROW_MAJOR_LAYOUT,
        device=device,
    )
    program_config = ttnn.SDPAProgramConfig(
        compute_with_storage_grid_size=ttnn.CoreCoord(
            device.compute_with_storage_grid_size().x,
            device.compute_with_storage_grid_size().y,
        ),
        q_chunk_size=32,
        k_chunk_size=k_chunk,
        exp_approx_mode=exp_approx,
    )
    out = ttnn.transformer.paged_scaled_dot_product_attention_decode(
        tt(q),
        tt(paged_layout(k, page_table, num_blocks)),
        tt(paged_layout(v, page_table, num_blocks)),
        page_table_tensor=tt(page_table, layout=False),
        cur_pos_tensor=tt(cur_pos, layout=False),
        scale=SCALE,
        program_config=program_config,
    )
    return ttnn.to_torch(out)[:, :, :n_q, :], q, k, v, cur_pos


def section_self_check(device):
    """MHA removes GQA mapping and q-head padding from the picture."""
    out, q, k, v, cp = build(
        device, 0, BATCH, NUM_KV_HEADS, NUM_KV_HEADS, lambda b: torch.randint(1, 256, (b,), dtype=torch.int32)
    )
    l2 = rel_l2(out, torch_reference(q, k, v, cp, SCALE))
    print(f"[1] self-check  MHA {NUM_KV_HEADS}q/{NUM_KV_HEADS}kv  rel_l2={l2:.6f}")
    if l2 > 0.05:
        raise SystemExit(f"self-check FAILED (rel_l2 {l2:.6f}): harness is wrong, not the op.")


def section_single_token(device):
    """cur_pos=0 -> one unmasked key -> output must equal V[0] bit-exactly."""
    print("\n[2] single unmasked token (cur_pos=0): output must be an exact copy of V[0]")
    print(
        f"    {'B':>3} {'Hq':>3} {'Hkv':>4} {'kchunk':>7} {'expapx':>7} "
        f"{'scale':>9} {'rel_l2':>9} {'rel_l2 /scale':>14} {'PCC':>9}"
    )
    zeros = lambda b: torch.zeros(b, dtype=torch.int32)  # noqa: E731
    cases = [
        (BATCH, NUM_Q_HEADS, NUM_KV_HEADS, 32, False),
        (BATCH, 32, 8, 32, False),
        (BATCH, 32, 32, 32, False),
        (1, 32, 8, 32, False),
        (BATCH, NUM_Q_HEADS, NUM_KV_HEADS, 64, False),
        (BATCH, NUM_Q_HEADS, NUM_KV_HEADS, 128, False),
        (BATCH, NUM_Q_HEADS, NUM_KV_HEADS, 32, True),
    ]
    for batch, n_q, n_kv, kc, apx in cases:
        out, _q, _k, v, _cp = build(device, 0, batch, n_q, n_kv, zeros, k_chunk=kc, exp_approx=apx)
        want = v[:, :, 0, :].repeat_interleave(n_q // n_kv, dim=1).unsqueeze(0).float()
        got = out.float()
        scale = ((got * want).sum() / (want * want).sum()).item()
        print(
            f"    {batch:3d} {n_q:3d} {n_kv:4d} {kc:7d} {str(apx):>7} "
            f"{scale:9.6f} {rel_l2(got, want):9.6f} {rel_l2(got / scale, want):14.6f} "
            f"{pcc(got, want):9.6f}"
        )
    print("    ^ scale is a hard constant; dividing it out leaves plain bf16 round-off.")


def section_seeds(device, seeds):
    print(f"\n[3] realistic decode positions, {seeds} seeds")
    print(f"    {'seed':>4} {'rel_l2':>10} {'PCC':>10}")
    pos_fn = lambda b: torch.randint(1, MAX_SEQ_LEN // 4, (b,), dtype=torch.int32)  # noqa: E731
    l2s = []
    for s in range(seeds):
        out, q, k, v, cp = build(device, s, BATCH, NUM_Q_HEADS, NUM_KV_HEADS, pos_fn)
        ref = torch_reference(q, k, v, cp, SCALE)
        l2s.append(rel_l2(out, ref))
        print(f"    {s:4d} {l2s[-1]:10.6f} {pcc(out, ref):10.6f}")
    l2s.sort()
    print(f"    median {l2s[len(l2s) // 2]:.6f}  min {l2s[0]:.6f}  max {l2s[-1]:.6f}")


def section_seq_len(device):
    print("\n[4] error vs sequence length (bf16 accumulation noise would grow as sqrt(N))")
    print(f"    {'seq':>6} {'rel_l2':>10}")
    for pos in (31, 63, 127, 255, 511, 1023):
        out, q, k, v, cp = build(
            device, 0, BATCH, NUM_Q_HEADS, NUM_KV_HEADS,
            lambda b, p=pos: torch.full((b,), p, dtype=torch.int32),
        )
        print(f"    {pos + 1:6d} {rel_l2(out, torch_reference(q, k, v, cp, SCALE)):10.6f}")
    print("    ^ flat, so this is not accumulation-bound.")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--seeds", type=int, default=10)
    ap.add_argument("--device-id", type=int, default=0)
    args = ap.parse_args()

    device = ttnn.open_device(device_id=args.device_id)
    try:
        g = device.compute_with_storage_grid_size()
        print(f"arch={device.arch()}  grid={g.x}x{g.y}")
        print(
            f"q=[1,{BATCH},{NUM_Q_HEADS},{HEAD_DIM}] bf16  "
            f"paged k/v=[{BATCH * BLOCKS_PER_USER},{NUM_KV_HEADS},{BLOCK_SIZE},{HEAD_DIM}] bf16  "
            f"scale={SCALE}  q_chunk=32\n"
        )
        section_self_check(device)
        section_single_token(device)
        section_seeds(device, args.seeds)
        section_seq_len(device)
    finally:
        ttnn.close_device(device)


if __name__ == "__main__":
    main()
```

</details>

**Comment by kmabeeTT** [comment]
## A debug patch we tried on forge side

Treat this as an exploratory artifact... I had Claude iterate on the kernel until the error in the repro test went away, so we could test whether fixing it helped the Falcon3-7B evals in [tt-inference-server#4752](https://github.com/tenstorrent/tt-inference-server/issues/4752). It didn't move the eval accuracy results, but the kernel numbers did change a lot, so it may be a useful starting point.

Branch: **[`kmabee/sdpa_accurate_exp`](https://github.com/tenstorrent/tt-metal/tree/kmabee/sdpa_accurate_exp)** — `b6371bb6488` on `f1f4ff75579`, one file, 8 lines.

### What it changes

`sub_exp_block_bcast_cols_inplace` in `.../sdpa/device/kernels/compute/compute_common.hpp` hardcodes `exp_tile<true /* approx */>`, so SDPA always uses the approximate exponential. `sdpa_decode_program_factory.cpp:767` does define `EXP_APPROX_MODE`, but the compute kernel never reads it — which may be why toggling `exp_approx_mode` gave you bit-identical output.

Forwarding the flag directly doesn't link: the accurate bf16 path folds the scale into `TTI_SFPMULI`, which needs a compile-time immediate. The `is_fp32_dest_acc_en` branch takes a runtime scale, so the patch uses the accurate path only when `fp32_dest_acc_en` is set.

Separately, the 0.9702 constant appears to be `math_approx_mode` (ttnn defaults it true, giving a reduced-refinement reciprocal) — no code change needed for that one, `compute_kernel_config` reaches it.

### Measured on P150, above repro unmodified

Single unmasked token, output should be a bit-exact copy of `V[0]`:

| config | scale | rel_l2 |
|:--|--:|--:|
| default | 0.970164 | 0.029920 |
| HiFi4 / approx=False / fp32=True, with patch | 1.000000 | 0.000000 |

Realistic decode positions:

| config | rel_l2 |
|:--|--:|
| default | 0.024995 |
| HiFi4/approx=False/fp32=True, before patch | 0.020278 |
| HiFi4/approx=False/fp32=True, with patch | 0.004872 |

For reference, a staged torch reference that rounds where the kernel rounds (scores and probs to bf16, bf16 reciprocal) sits at 0.003512.

### Caveats

Throughput is not measured — `exp` is the inner loop and this header is shared with prefill, so the `fp32_dest_acc_en` gating is probably not the right long-term shape. And as above, it did not improve the evals this came from, so there's no accuracy-at-the-model-level argument behind it.

Data, sweeps and scripts: **https://gist.github.com/kmabeeTT/24f5ff1c2f0b957008ea8e7db494d740**

**Comment by dddrrreee** [comment]
This hit me, really burned a lot of time and tokens.    Astra/fable narrowed down, astra made a patch, another astra+fable replicated etc.   LMK if operator error.


 Attached: a minimal reproducer, proposed two-line correction, and independent
 Blackhole hardware validation. The report documents the ignored
 exp_approx_mode=False setting, measured intermediate values, contributor credits,
 and the tested scope of the correction.

[tt-sdpa-accurate-exp-reproducer-20260918.zip](https://github.com/user-attachments/files/32391762/tt-sdpa-accurate-exp-reproducer-20260918.zip)
